src/ERC4626StreamHub.sol

```diff
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.10;
+
+import "forge-std/console2.sol";
+
+import {IERC4626} from "openzeppelin-contracts/interfaces/IERC4626.sol";
+import {IERC20} from "openzeppelin-contracts/interfaces/IERC20.sol";
+import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol";
+import {Multicall} from "openzeppelin-contracts/utils/Multicall.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+
+/**
+ * @title ERC4626StreamHub
+ * @dev This contract implements a stream hub for managing yield streams between senders and receivers.
+ * It allows users to open yield streams, claim yield from streams, and close streams to withdraw remaining shares.
+ * The contract uses the ERC4626 interface for interacting with the underlying vault.
+ */
+contract ERC4626StreamHub is Multicall {
+    using FixedPointMathLib for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for IERC4626;
+
+    error NotEnoughShares();
+    error ZeroShares();
+    error AddressZero();
+    error CannotOpenStreamToSelf();
+    error StreamDoesNotExist();
+    error NoYieldToClaim();
+    error InputParamsLengthMismatch();
+
+    event Deposit(address indexed depositor, uint256 shares);
+    event Withdraw(address indexed depositor, uint256 shares);
+    event OpenYieldStream(
+        address indexed streamer,
+        address indexed receiver,
+        uint256 shares,
+        uint256 principal
+    );
+    event ClaimYield(
+        address indexed receiver,
+        address indexed claimedTo,
+        uint256 yield
+    );
+    event CloseYieldStream(
+        address indexed streamer,
+        address indexed receiver,
+        uint256 shares
+    );
+
+    IERC4626 public immutable vault;
+
+    // receiver to number of shares it is entitled to as the yield beneficiary
+    mapping(address => uint256) public receiverShares;
+
+    // receiver to total amount of assets (principal) - not claimable
+    mapping(address => uint256) public receiverTotalPrincipal;
+
+    // receiver to total amount of assets (principal) allocated from a single address
+    mapping(address => mapping(address => uint256)) public receiverPrincipal;
+
+    constructor(IERC4626 _vault) {
+        vault = _vault;
+        IERC20(vault.asset()).safeApprove(address(vault), type(uint256).max);
+    }
+
+    /**
+     * @dev Opens a yield stream for a specific receiver with a given number of shares.
+     * @param _receiver The address of the receiver.
+     * @param _shares The number of shares to allocate for the yield stream.
+     */
+    function openYieldStream(address _receiver, uint256 _shares) public {
```

Comment:
also opening with a permit instead of approve would be a nice addition

Reply:
[added](https://github.com/lindy-labs/erc4626-utils/pull/1/commits/fdba9fe90df3f9b72030319e82c00a01dc543874)